src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -380,11 +380,17 @@ abstract class RefChecks extends Transform {
           // todo: align accessibility implication checking with isAccessible in Contexts
           val ob = other.accessBoundary(memberClass)
           val mb = member.accessBoundary(memberClass)
-          def isOverrideAccessOK = member.isPublic || {      // member is public, definitely same or relaxed access
-            (!other.isProtected || member.isProtected) &&    // if o is protected, so is m
-            ((!isRootOrNone(ob) && ob.hasTransOwner(mb)) ||  // m relaxes o's access boundary
-             (other.isJavaDefined && other.isProtected))     // overriding a protected java member, see #3946 #12349
-          }
+          @inline def protectedOK = !other.isProtected || member.isProtected
+          @inline def companionBoundaryOK = ob.isClass && mb.isModuleClass && (mb.module == ob.companionSymbol ||
+            !ob.companionSymbol.exists && ob.isLocalToBlock && ob.owner == mb.owner && ob.name.toTermName == mb.module.name)
```

Comment:
This is wrong, because templates in different scopes can have the same owner, but they are not companions.

Reply:
Has anyone tried to dig into the reason why local companionship doesn't work out?

Reply:
@joroKr21 maybe I'll employ the "sunk-cost fallacy" to pursue the question, while my limited knowledge is still fresh, and while my time is unstructured. The angry code comment is quite old. I see the lrytz fix from a decade ago, but the statute of limitations has run out. It seems to come up every half-decade, because then retronym https://github.com/scala/scala/pull/5550#issue-190710957 I'll start there; as I commented in 2016, "I've forgotten the context."  I don't think Refchecks has the contexts for this lookup; at the least, typer could notice at the end of a block that a companion look-up failed because of a forward reference.